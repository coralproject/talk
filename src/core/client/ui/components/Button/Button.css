.root {
  composes: button from "talk-ui/shared/typography.css";
  padding: 5px 20px;
  border-radius: $round-corners;
  background-color: transparent;
  /* TODO: hover styles for the default button */
}

.fullWidth {
  display: block;
  width: 100%;
  box-sizing: border-box;
}

.primary {
  background-color: $palette-primary-main;
  color: #fff;

  &:hover {
    background-color: $palette-primary-light;
  }

  &:active {
    background-color: $palette-primary-lighter;
  }

  &.invert {
    background-color: transparent;
    border: 1px solid $palette-primary-main;
    color: $palette-primary-main;

    &:hover {
      border-color: $palette-primary-light;
      color: $palette-primary-light;
    }

    &:active {
      border-color: $palette-primary-lighter;
      color: $palette-primary-lighter;
    }
  }
}

.secondary {
  background-color: $palette-secondary-main;
  color: #fff;

  &:hover {
    background-color: $palette-secondary-light;
  }

  &:active {
    background-color: $palette-secondary-lighter;
  }

  &.invert {
    background-color: transparent;
    border: 1px solid $palette-secondary-main;
    color: $palette-secondary-main;

    &:hover {
      border-color: $palette-secondary-light;
      color: $palette-secondary-light;
    }

    &:active {
      border-color: $palette-secondary-lighter;
      color: $palette-secondary-lighter;
    }
  }
}

/**
 * This seems to be the best way to target modern touch device browsers.
 */
@media (-moz-touch-enabled: 1), (pointer:coarse) {
  /* TODO: Remove hover styles */
}
