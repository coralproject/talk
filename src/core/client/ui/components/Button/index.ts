export * from "./Button";
export { default } from "./Button";
