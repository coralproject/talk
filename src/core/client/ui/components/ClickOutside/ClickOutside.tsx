import React from "react";
import { findDOMNode } from "react-dom";

interface Props {
  onClickOutside: () => void;
  children: React.ReactNode;
}

class ClickOutside extends React.Component<Props> {
  public domNode: Element | null = null;

  public handleClick = (e: MouseEvent) => {
    const { onClickOutside } = this.props;
    if (!e || !this.domNode!.contains(e.target as HTMLInputElement)) {
      // tslint:disable-next-line:no-unused-expression
      onClickOutside && onClickOutside();
    }
  };

  public componentDidMount() {
    this.domNode = findDOMNode(this) as Element;
    document.addEventListener("click", this.handleClick, true);
  }

  public componentWillUnmount() {
    document.removeEventListener("click", this.handleClick, true);
  }

  public render() {
    return this.props.children;
  }
}
export default ClickOutside;
