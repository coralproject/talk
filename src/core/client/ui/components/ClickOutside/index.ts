export { default as ClickOutside, ClickFarAwayRegister } from "./ClickOutside";
