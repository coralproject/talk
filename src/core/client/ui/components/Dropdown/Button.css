.root {
  composes: menuItem from "coral-ui/shared/typography.css";
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: var(--spacing-2);
  width: 100%;
  text-align: left;
  line-height: 1;
}

.anchor {
  color: var(--palette-primary-main);
  text-decoration: underline;
  &:visited {
    color: var(--palette-primary-main);
  }
}

.root:disabled {
  opacity: 0.6;
}

.root:not(:disabled):active {
  background-color: var(--palette-primary-lightest);
}

.mouseHover {
  background-color: var(--palette-primary-lightest);
}

.iconBefore {
  display: flex;
  justify-content: baseline;
  align-items: center;
  color: var(--palette-grey-main);
  margin-right: var(--spacing-2);
}
.iconAfter {
  display: flex;
  align-items: center;
  justify-content: baseline;
  margin-left: var(--spacing-2);
  text-decoration: unset;
}
.iconOpenInNew {
  padding-top: 1px;
  color: var(--palette-primary-main);
}

.blankAdornment {
  padding-right: var(--spacing-6);
}
