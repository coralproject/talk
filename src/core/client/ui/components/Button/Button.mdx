---
name: Button
menu: UI Kit
---

import { Playground } from 'docz'
import Button from './Button'
import ButtonIcon from './ButtonIcon'
import Icon from '../Icon'
import Flex from '../Flex'

# Button

## Regular Button
<Playground>
  <Flex itemGutter>
    <Button>Push Me</Button>
    <Button size="small">Push Me</Button>
    <Button size="large">Push Me</Button>
    <Button color="primary">Push Me</Button>
    <Button color="error">Push Me</Button>
    <Button color="success">Push Me</Button>
    <Button disabled>Push Me</Button>
    <Button active>Push Me</Button>
    <Button><ButtonIcon>face</ButtonIcon><span>Push Me</span></Button>
    <Button fullWidth>Push Me</Button>
  </Flex>
</Playground>

## Filled Button
<Playground>
  <Flex itemGutter>
    <Button variant="filled">Push Me</Button>
    <Button variant="filled" size="small">Push Me</Button>
    <Button variant="filled" size="large">Push Me</Button>
    <Button variant="filled" color="primary">Push Me</Button>
    <Button variant="filled" color="error">Push Me</Button>
    <Button variant="filled" color="success">Push Me</Button>
    <Button variant="filled" disabled>Push Me</Button>
    <Button variant="filled" active>Push Me</Button>
    <Button variant="filled"><ButtonIcon>face</ButtonIcon><span>Push Me</span></Button>
    <Button variant="filled" fullWidth>Push Me</Button>
  </Flex>
</Playground>

## Outlined Button
<Playground>
  <Flex itemGutter>
    <Button variant="outlined">Push Me</Button>
    <Button variant="outlined" size="small">Push Me</Button>
    <Button variant="outlined" size="large">Push Me</Button>
    <Button variant="outlined" color="primary">Push Me</Button>
    <Button variant="outlined" color="error">Push Me</Button>
    <Button variant="outlined" color="success">Push Me</Button>
    <Button variant="outlined" disabled>Push Me</Button>
    <Button variant="outlined" active>Push Me</Button>
    <Button variant="outlined"><ButtonIcon>face</ButtonIcon><span>Push Me</span></Button>
    <Button variant="outlined" fullWidth>Push Me</Button>
  </Flex>
</Playground>

## Ghost Button
<Playground>
  <Flex itemGutter>
    <Button variant="ghost">Push Me</Button>
    <Button variant="ghost" size="small">Push Me</Button>
    <Button variant="ghost" size="large">Push Me</Button>
    <Button variant="ghost" color="primary">Push Me</Button>
    <Button variant="ghost" color="error">Push Me</Button>
    <Button variant="ghost" color="success">Push Me</Button>
    <Button variant="ghost" disabled>Push Me</Button>
    <Button variant="ghost" active>Push Me</Button>
    <Button variant="ghost"><ButtonIcon>face</ButtonIcon><span>Push Me</span></Button>
    <Button variant="ghost" fullWidth>Push Me</Button>
  </Flex>
</Playground>
