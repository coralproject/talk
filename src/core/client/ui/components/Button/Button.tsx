import cn from "classnames";
import { pick } from "lodash";
import React, { ButtonHTMLAttributes, Ref } from "react";

import { withForwardRef, withStyles } from "talk-ui/hocs";
import { PropTypesOf } from "talk-ui/types";

import BaseButton, { BaseButtonProps } from "../BaseButton";
import * as styles from "./Button.css";

// This should extend from BaseButton instead but we can't because of this bug
// TODO: add bug link.
interface InnerProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** If set renders an anchor tag instead */
  anchor?: boolean;
  href?: string;
  target?: string;
  /**
   * This prop can be used to add custom classnames.
   * It is handled by the `withStyles `HOC.
   */
  classes: typeof styles & BaseButtonProps["classes"];

  /** Size of the button */
  size?: "small" | "regular" | "large";

  /** Color of the button */
  color?: "regular" | "primary" | "error" | "success";

  /** Variant of the button */
  variant?: "regular" | "filled" | "outlined" | "ghost" | "underlined";

  /** If set renders a full width button */
  fullWidth?: boolean;

  /** If set renders active state e.g. to implement toggle buttons */
  active?: boolean;

  type?: "submit" | "reset" | "button";

  /** Internal: Forwarded Ref */
  forwardRef?: Ref<HTMLButtonElement>;
}

export class Button extends React.Component<InnerProps> {
  public static defaultProps: Partial<InnerProps> = {
    size: "regular",
    variant: "regular",
    color: "regular",
  };
  public render() {
    const {
      active,
      classes,
      color,
      className,
      size,
      fullWidth,
      disabled,
      forwardRef,
      variant,
      type,
      ...rest
    } = this.props;

    const rootClassName = cn(classes.root, className, {
      [classes.sizeRegular]: size === "regular",
      [classes.sizeSmall]: size === "small",
      [classes.sizeLarge]: size === "large",
      [classes.colorRegular]: color === "regular",
      [classes.colorPrimary]: color === "primary",
      [classes.colorError]: color === "error",
      [classes.colorSuccess]: color === "success",
      [classes.variantRegular]: variant === "regular",
      [classes.variantFilled]: variant === "filled",
      [classes.variantOutlined]: variant === "outlined",
      [classes.variantGhost]: variant === "ghost",
      [classes.variantUnderlined]: variant === "underlined",
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.disabled]: disabled,
    });

    return (
      <BaseButton
        className={rootClassName}
        classes={pick(classes, "keyboardFocus", "mouseHover")}
        disabled={disabled}
        forwardRef={forwardRef}
        type={type}
        {...rest}
      />
    );
  }
}

const enhanced = withForwardRef(withStyles(styles)(Button));
export type ButtonProps = PropTypesOf<typeof enhanced>;
export default enhanced;
