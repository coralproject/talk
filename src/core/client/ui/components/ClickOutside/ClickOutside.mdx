---
name: ClickOutside
menu: UI Kit
---

import { Playground, PropsTable } from 'docz'
import ClickOutside from './ClickOutside'
import Button from '../Button/Button'

# ClickOutside

A Component to handle click events outside the children components.

## Basic usage
Wrap a Component with `<ClickOutside>` and pass a function to `onClickOutside`. This function will trigger when the user
clicks outside the component.

### Test
Click the blue background. It should trigger an alert. Nothing should happen if you click the button.

<Playground>
  <div style={{background: 'blue', padding: '10px'}}>
    <ClickOutside onClickOutside={() => alert('You clicked outside!')}>
      <Button variant="filled">Push Me</Button>
    </ClickOutside>
  </div>
</Playground>
