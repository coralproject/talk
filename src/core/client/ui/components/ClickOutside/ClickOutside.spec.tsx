import { mount } from "enzyme";
import React from "react";
import simulant from "simulant";
import sinon from "sinon";

import Button from "../Button/Button";
import ClickOutside from "./ClickOutside";

let container: HTMLElement;

beforeAll(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterAll(() => {
  document.body.removeChild(container);
});

it("should detect click outside", () => {
  const onClickOutside = sinon.spy();
  const wrapper = mount(
    <ClickOutside onClickOutside={onClickOutside}>
      <Button variant="filled">Push Me</Button>
    </ClickOutside>,
    {
      attachTo: container,
    }
  );
  simulant.fire(container, "click");

  expect(onClickOutside.calledOnce).toEqual(true);
  expect(wrapper).toMatchSnapshot();
  wrapper.unmount();
});

it("should ignore click inside", () => {
  const onClickOutside = sinon.spy();
  const wrapper = mount(
    <ClickOutside onClickOutside={onClickOutside}>
      <Button id="click-outside-test-button" variant="filled">
        Push Me
      </Button>
    </ClickOutside>,
    {
      attachTo: container,
    }
  );
  simulant.fire(document.getElementById("click-outside-test-button")!, "click");

  expect(onClickOutside.calledOnce).toEqual(false);
  expect(wrapper).toMatchSnapshot();
  wrapper.unmount();
});
