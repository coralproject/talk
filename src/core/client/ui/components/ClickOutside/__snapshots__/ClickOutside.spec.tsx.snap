// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should detect click outside 1`] = `
<ClickOutside
  onClickOutside={[Function]}
>
  <withPropsOnChange(Button)
    variant="filled"
  >
    <Button
      classes={
        Object {
          "active": "Button-active",
          "colorError": "Button-colorError",
          "colorPrimary": "Button-colorPrimary",
          "colorRegular": "Button-colorRegular",
          "colorSuccess": "Button-colorSuccess",
          "default": Object {
            "active": "Button-active",
            "colorError": "Button-colorError",
            "colorPrimary": "Button-colorPrimary",
            "colorRegular": "Button-colorRegular",
            "colorSuccess": "Button-colorSuccess",
            "disabled": "Button-disabled",
            "disabled)": "Button-disabled)",
            "fullWidth": "Button-fullWidth",
            "mouseHover": "Button-mouseHover",
            "root": "Button-root",
            "sizeLarge": "Button-sizeLarge",
            "sizeRegular": "Button-sizeRegular",
            "sizeSmall": "Button-sizeSmall",
            "variantFilled": "Button-variantFilled",
            "variantGhost": "Button-variantGhost",
            "variantOutlined": "Button-variantOutlined",
            "variantRegular": "Button-variantRegular",
          },
          "disabled": "Button-disabled",
          "disabled)": "Button-disabled)",
          "fullWidth": "Button-fullWidth",
          "mouseHover": "Button-mouseHover",
          "root": "Button-root",
          "sizeLarge": "Button-sizeLarge",
          "sizeRegular": "Button-sizeRegular",
          "sizeSmall": "Button-sizeSmall",
          "variantFilled": "Button-variantFilled",
          "variantGhost": "Button-variantGhost",
          "variantOutlined": "Button-variantOutlined",
          "variantRegular": "Button-variantRegular",
        }
      }
      color="regular"
      size="regular"
      variant="filled"
    >
      <withPropsOnChange(WithMouseHover)
        className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
        classes={
          Object {
            "mouseHover": "Button-mouseHover",
          }
        }
      >
        <WithMouseHover
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
          classes={
            Object {
              "default": Object {
                "keyboardFocus": "BaseButton-keyboardFocus",
                "mouseHover": "BaseButton-mouseHover",
                "root": "BaseButton-root",
              },
              "keyboardFocus": "BaseButton-keyboardFocus",
              "mouseHover": "BaseButton-mouseHover Button-mouseHover",
              "root": "BaseButton-root",
            }
          }
        >
          <WithKeyboardFocus
            className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
            classes={
              Object {
                "default": Object {
                  "keyboardFocus": "BaseButton-keyboardFocus",
                  "mouseHover": "BaseButton-mouseHover",
                  "root": "BaseButton-root",
                },
                "keyboardFocus": "BaseButton-keyboardFocus",
                "mouseHover": "BaseButton-mouseHover Button-mouseHover",
                "root": "BaseButton-root",
              }
            }
            mouseHover={false}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
          >
            <BaseButton
              className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
              classes={
                Object {
                  "default": Object {
                    "keyboardFocus": "BaseButton-keyboardFocus",
                    "mouseHover": "BaseButton-mouseHover",
                    "root": "BaseButton-root",
                  },
                  "keyboardFocus": "BaseButton-keyboardFocus",
                  "mouseHover": "BaseButton-mouseHover Button-mouseHover",
                  "root": "BaseButton-root",
                }
              }
              keyboardFocus={false}
              mouseHover={false}
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
            >
              <button
                className="BaseButton-root Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
              >
                Push Me
              </button>
            </BaseButton>
          </WithKeyboardFocus>
        </WithMouseHover>
      </withPropsOnChange(WithMouseHover)>
    </Button>
  </withPropsOnChange(Button)>
</ClickOutside>
`;

exports[`should ignore click inside 1`] = `
<ClickOutside
  onClickOutside={[Function]}
>
  <withPropsOnChange(Button)
    id="click-outside-test-button"
    variant="filled"
  >
    <Button
      classes={
        Object {
          "active": "Button-active",
          "colorError": "Button-colorError",
          "colorPrimary": "Button-colorPrimary",
          "colorRegular": "Button-colorRegular",
          "colorSuccess": "Button-colorSuccess",
          "default": Object {
            "active": "Button-active",
            "colorError": "Button-colorError",
            "colorPrimary": "Button-colorPrimary",
            "colorRegular": "Button-colorRegular",
            "colorSuccess": "Button-colorSuccess",
            "disabled": "Button-disabled",
            "disabled)": "Button-disabled)",
            "fullWidth": "Button-fullWidth",
            "mouseHover": "Button-mouseHover",
            "root": "Button-root",
            "sizeLarge": "Button-sizeLarge",
            "sizeRegular": "Button-sizeRegular",
            "sizeSmall": "Button-sizeSmall",
            "variantFilled": "Button-variantFilled",
            "variantGhost": "Button-variantGhost",
            "variantOutlined": "Button-variantOutlined",
            "variantRegular": "Button-variantRegular",
          },
          "disabled": "Button-disabled",
          "disabled)": "Button-disabled)",
          "fullWidth": "Button-fullWidth",
          "mouseHover": "Button-mouseHover",
          "root": "Button-root",
          "sizeLarge": "Button-sizeLarge",
          "sizeRegular": "Button-sizeRegular",
          "sizeSmall": "Button-sizeSmall",
          "variantFilled": "Button-variantFilled",
          "variantGhost": "Button-variantGhost",
          "variantOutlined": "Button-variantOutlined",
          "variantRegular": "Button-variantRegular",
        }
      }
      color="regular"
      id="click-outside-test-button"
      size="regular"
      variant="filled"
    >
      <withPropsOnChange(WithMouseHover)
        className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
        classes={
          Object {
            "mouseHover": "Button-mouseHover",
          }
        }
        id="click-outside-test-button"
      >
        <WithMouseHover
          className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
          classes={
            Object {
              "default": Object {
                "keyboardFocus": "BaseButton-keyboardFocus",
                "mouseHover": "BaseButton-mouseHover",
                "root": "BaseButton-root",
              },
              "keyboardFocus": "BaseButton-keyboardFocus",
              "mouseHover": "BaseButton-mouseHover Button-mouseHover",
              "root": "BaseButton-root",
            }
          }
          id="click-outside-test-button"
        >
          <WithKeyboardFocus
            className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
            classes={
              Object {
                "default": Object {
                  "keyboardFocus": "BaseButton-keyboardFocus",
                  "mouseHover": "BaseButton-mouseHover",
                  "root": "BaseButton-root",
                },
                "keyboardFocus": "BaseButton-keyboardFocus",
                "mouseHover": "BaseButton-mouseHover Button-mouseHover",
                "root": "BaseButton-root",
              }
            }
            id="click-outside-test-button"
            mouseHover={false}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            onTouchEnd={[Function]}
          >
            <BaseButton
              className="Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
              classes={
                Object {
                  "default": Object {
                    "keyboardFocus": "BaseButton-keyboardFocus",
                    "mouseHover": "BaseButton-mouseHover",
                    "root": "BaseButton-root",
                  },
                  "keyboardFocus": "BaseButton-keyboardFocus",
                  "mouseHover": "BaseButton-mouseHover Button-mouseHover",
                  "root": "BaseButton-root",
                }
              }
              id="click-outside-test-button"
              keyboardFocus={false}
              mouseHover={false}
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
            >
              <button
                className="BaseButton-root Button-root Button-sizeRegular Button-colorRegular Button-variantFilled"
                id="click-outside-test-button"
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseDown={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
              >
                Push Me
              </button>
            </BaseButton>
          </WithKeyboardFocus>
        </WithMouseHover>
      </withPropsOnChange(WithMouseHover)>
    </Button>
  </withPropsOnChange(Button)>
</ClickOutside>
`;
